### PR TITLE
Add wait_for method and node event client

### PR DESCRIPTION
**Pull request by leebradley**
Add wait_for convenience method, which allows a node to wait for a
particular admin event type rather than manually looping through every
event.

Add persistent node event client for nodes that will listen for all
event types and track the current event state. This event client is
currently only used by the wait_for convenience method.

Signed-off-by: Lee Bradley <bradley@bitwise.io>

### CODE REVIEW

File: libsplinter/src/admin/client/event/mod.rs [hunk 1]
```diff
@@ -93,6 +94,15 @@ pub trait AdminServiceEventClient {
     ///
     /// This should block until an event is available.
     fn next_event(&self) -> Result<AdminServiceEvent, NextEventError>;
+
+    /// Wait for the given event up until a timeout
+    ///
+    /// This should block until an event is available.
+    fn wait_for(
+        &self,
+        event_type: &EventType,
```
Comment (peterschwarz) [inline libsplinter/src/admin/client/event/mod.rs:103]: This argument needs to contain more detail then just the type.  More likely then not, the test will be waiting on a specific circuit change.

Perhaps an enum like
```rust
enum EventQuery {
    CircuitReady(String), // or a struct variant with circuit id
    CircuitRejected(String),
    CircuitDisbanded(String),
    // ... etc
    // The voting ones would be similar but include the PublicKey
}
```

File: libsplinter/src/admin/client/event/ws/actix_web_client.rs [hunk 2]
```diff
@@ -303,6 +307,32 @@ impl AdminServiceEventClient for AwcAdminServiceEventClient {
         evt_result
             .map_err(|e| NextEventError::InternalError(InternalError::from_source(Box::new(e))))
     }
+
+    fn wait_for(
```
Comment (peterschwarz) [inline libsplinter/src/admin/client/event/ws/actix_web_client.rs:311]: This function is implementation-agnostic, so it can be pulled up to the trait as a default implementation. Any implementer that wants to override it can do so.

**Review by peterschwarz — CHANGES_REQUESTED** [review]

File: libsplinter/src/admin/client/event/mod.rs (hunk 1 above)
Comment (leebradley) [reply libsplinter/src/admin/client/event/mod.rs]: That's a good point.

I created an `EventQuery` enum that's consumed by `wait_for` and uses the filters you list above. To cover all possible uses cases, `wait_for` calls a new function `wait_for_filter` that will accept any arbitrary closure to act as an event filter.

Under the hood, the `EventQuery` enum just transforms itself into an event filter closure.

File: libsplinter/src/admin/client/event/mod.rs [hunk 3]
```diff
@@ -79,6 +82,54 @@ impl AdminServiceEvent {
...
+    ProposalVote { circuit_id: &'a str, key: PublicKey },
+    ProposalAccepted { circuit_id: &'a str, key: PublicKey },
+    ProposalRejected { circuit_id: &'a str, key: PublicKey },
+    CircuitReady { circuit_id: &'a str },
+    CircuitDisbanded { circuit_id: &'a str },
+}
+
+impl<'a> EventQuery<'a> {
+    /// Transform the EventQuery into an event filter
+    pub fn filter(self) -> impl Fn(&AdminServiceEvent) -> bool + 'a {
```
Comment (peterschwarz) [inline libsplinter/src/admin/client/event/mod.rs:97]: This probably doesn't need to be public

File: libsplinter/src/admin/client/event/ws/actix_web_client.rs [hunk 4]
```diff
@@ -159,6 +159,7 @@ impl RunnableAwcAdminServiceEventClient {
         let last_event_id = Arc::new(AtomicU64::new(last_event_id.unwrap_or(0)));
         let received_id = last_event_id.clone();
         let received_sender = event_sender.clone();
+
```
Comment (peterschwarz) [inline libsplinter/src/admin/client/event/ws/actix_web_client.rs:162]: As this is formatting related it should be part of a separate commit (or the change should be removed)

**Review by peterschwarz — DISMISSED** [review]

**Review by shannynalayna — DISMISSED** [review]

**Review by shannynalayna — APPROVED** [review]

**Review by peterschwarz — APPROVED** [review]
